mapeditor/inspector/townevent.cpp

```diff
@@ -0,0 +1,320 @@
+/*
+ * townevent.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "../StdInc.h"
+#include "townbuildingswidget.h"
+#include "townevent.h"
+#include "ui_townevent.h"
+#include "../../lib/constants/NumericConstants.h"
+#include "../../lib/constants/StringConstants.h"
+
+TownEvent::TownEvent(CGTownInstance & t, QListWidgetItem * item, QWidget * parent) :
+	QDialog(parent),
+	ui(new Ui::TownEvent),
+	town(t),
+	item(item)
+{
+	ui->setupUi(this);
+
+	ui->buildingsTree->setModel(&buildingsModel);
+
+	params = item->data(Qt::UserRole).toMap();
+	ui->eventFirstOccurrence->setMinimum(1);
+	ui->eventFirstOccurrence->setMaximum(999);
+	ui->eventRepeatAfter->setMaximum(999);
+	ui->eventNameText->setText(params.value("name").toString());
+	ui->eventMessageText->setPlainText(params.value("message").toString());
+	ui->eventAffectsCpu->setChecked(params.value("computerAffected").toBool());
+	ui->eventAffectsHuman->setChecked(params.value("humanAffected").toBool());
+	ui->eventFirstOccurrence->setValue(params.value("firstOccurrence").toInt()+1);
+	ui->eventRepeatAfter->setValue(params.value("nextOccurrence").toInt());
+
+	initPlayers();
+	initResources();
+	initBuildings();
+	initCreatures();
+
+	show();
+}
+
+TownEvent::~TownEvent()
+{
+	delete ui;
+}
+
+void TownEvent::initPlayers()
+{
+	for (int i = 0; i < PlayerColor::PLAYER_LIMIT_I; ++i)
+	{
+		bool isAffected = (1 << i) & params.value("players").toInt();
+		auto * item = new QListWidgetItem(QString::fromStdString(GameConstants::PLAYER_COLOR_NAMES[i]));
+		item->setData(Qt::UserRole, QVariant::fromValue(i));
+		item->setCheckState(isAffected ? Qt::Checked : Qt::Unchecked);
+		ui->playersAffected->addItem(item);
+	}
+}
+
+void TownEvent::initResources()
+{
+	ui->resourcesTable->setRowCount(GameConstants::RESOURCE_QUANTITY);
+	auto resourcesMap = params.value("resources").toMap();
+	for (int i = 0; i < GameConstants::RESOURCE_QUANTITY; ++i)
+	{
+		auto name = QString::fromStdString(GameConstants::RESOURCE_NAMES[i]);
+		int val = resourcesMap.value(name).toInt();
+		ui->resourcesTable->setItem(i, 0, new QTableWidgetItem(name));
+
+		QSpinBox * edit = new QSpinBox(ui->resourcesTable);
+		edit->setMaximum(i == GameResID::GOLD ? 999999 : 999);
+		edit->setMinimum(i == GameResID::GOLD ? -999999 : -999);
```

Comment:
Why limit to 999?

Reply:
couldn't find any RESOURCE_LIMIT const so I used some numbers that other rewards widget is using

Reply:
magic numbers should be named constants